rustler_mix/lib/rustler/compiler.ex

```diff
@@ -69,46 +70,58 @@ defmodule Rustler.Compiler do
     ["rustup", "run", version, "cargo", "rustc"]
   end
 
-  defp make_platform_hacks(args, crate_path, {:unix, :darwin}) do
-    root = Path.join([".cargo", "config"])
-    path = Path.join([crate_path, ".cargo", "config"])
-
-    if File.exists?(root) || File.exists?(path) do
-      args
-    else
-      IO.write([
-        "\n",
-        IO.ANSI.yellow(),
-        """
-        Compiling on macOS requires special link args in order to compile
-        correctly.
-
-        Rustler is currently working around this issue in the compiler task.
-        This will be removed in v1.0.0 in favor of a user supplied .cargo/config
-        file.
-
-        To remove this warning, please create #{path}
-        with the following content:
-
-              [target.x86_64-apple-darwin]
-              rustflags = [
-                  "-C", "link-arg=-undefined",
-                  "-C", "link-arg=dynamic_lookup",
-              ]
-
-        See https://developer.apple.com/library/archive/documentation/DeveloperTools/Conceptual/MachOTopics/1-Articles/executing_files.html
-        for more details.
-
-        """,
-        IO.ANSI.default_color(),
-        "\n"
-      ])
-
-      args ++ ["--", "-C", "link-arg=-undefined", "-C", "link-arg=dynamic_lookup"]
+  defp ensure_platform_requirements!(crate_path, config, {:unix, :darwin}) do
+    workspace_root = config.metadata["workspace_root"]
+
+    components =
+      crate_path
+      |> Path.relative_to(workspace_root)
+      |> Path.split()
+
+    {potential_config_files, _} =
+      Enum.map_reduce(["" | components], workspace_root, fn component, path ->
+        path = Path.join(path, component)
+        file = Path.join([path, ".cargo", "config"])
+        {file, path}
+      end)
+
+    sorted_config_files =
+      potential_config_files
+      |> Enum.filter(&File.exists?/1)
+      |> Enum.reverse()
+
+    has_macos_target_os_configuration? =
+      sorted_config_files
+      |> Stream.map(&Toml.decode_file!/1)
+      |> Enum.drop_while(fn toml ->
+        toml
+        |> Map.get("target", [])
+        |> Enum.filter(fn {key, _} ->
+          String.match?(key, ~r/(.*macos.*)|(.*darwin.*)/)
+        end)
+      end)
+
+    unless has_macos_target_os_configuration? do
+      raise """
+      Compiling on macOS requires special link args in order to compile
+      correctly.
+
+      To remove this error, please create .cargo/config
+      with the following content:
+
+            [target.x86_64-apple-darwin]
```

Comment:
I think this suggestion can use the `cfg` for capturing the OS as you did in the other PR: https://github.com/rusterlium/rustler/pull/438/files#diff-aea852ba0a1ff20d4a5eab3df8c59e3fc4e91560f8a0e07d131e1ca2ca600127R93

```suggestion
            [target.'cfg(target_os = "macos")']
```

Reply:
Yeah, this conflicted with #438, I will rebase on that.